### ISSUE TITLE
git contention leads applications into Unknown state

### ISSUE THREAD

**Issue by phsiao** [argoproj/argo-cd#2351]
**Describe the bug**

When multiple applications sharing the same repo, on push now I frequently see git lock gets in the way of sync.

**To Reproduce**

Set multiple applications to share the same repo, in our case, we use kustomize and use different overlay layers for different clusters, and each application maps to a cluster.  Make a change to the repo, and multiple applications will start refresh and sync, and get into "Unknown" state.

If may take a while for all of them to resolve by themselves, and often require some manual refresh.

**Expected behavior**

Applications should not get into unknown state.

**Version**

v1.2.1

**Logs**

```git checkout --force 19f55d788e7410895a34ad443abb105e71a41975` failed exit status 128: fatal: Unable to create '/tmp/<repo name>/.git/index.lock': File exists. Another git process seems to be running in this repository, e.g. an editor opened by 'git commit'. Please make sure all processes are terminated then try again. If it still fails, a git process may have crashed in this repository earlier: remove the file manually to continue.```

**Comment by alexec** [comment]
This looks like a bug. The code for syncing has substantially been re-written for v1.3, so any bug fix for your issue might no longer apply.

**Comment by cmur2** [comment]
I encountered the same error with ArgoCD v1.3.0, from the repo server logs:
```
time="2019-11-26T15:29:27Z" level=error msg="finished unary call with code Internal" error="rpc error: code = Internal desc = Failed to checkout 771cf3365c74c27c9ce8a7b9b0e27ee450d5091d: `git checkout --force 771cf3365c74c27c9ce8a7b9b0e27ee450d5091d` failed exit status 128: fatal: Unable to create '/tmp/<repo name>/.git/index.lock': File exists.\n\nAnother git process seems to be running in this repository, e.g.\nan editor opened by 'git commit'. Please make sure all processes\nare terminated then try again. If it still fails, a git process\nmay have crashed in this repository earlier:\nremove the file manually to continue." grpc.code=Internal grpc.method=GetRevisionMetadata grpc.request.deadline="2019-11-26T15:30:25Z" grpc.service=repository.RepoServerService grpc.start_time="2019-11-26T15:29:26Z" grpc.time_ms=1111.512 span.kind=server system=grpc
```

We also use kustomize to bootstrap multiple clusters from the same repo so on push they trigger refresh nearly concurrently.

**Comment by alexec** [comment]
I believe this is probably fixed, or at least partially addressed, by #2771.

**Comment by alexmt** [comment]
Send PR with the fix: https://github.com/argoproj/argo-cd/pull/2877
